### PR TITLE
ddl: fix concurrent access to `do.infoHandle` has a data race error

### PR DESCRIPTION
**Pull request by ciscoxll**
### What problem does this PR solve? 

```go
WARNING: DATA RACE
Write at 0x00c0001d4c70 by goroutine 51:
  github.com/pingcap/tidb/ddl_test.(*testFailDBSuite).TestHalfwayCancelOperations()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/domain/domain.go:538 +0x1c80
  github.com/pingcap/tidb/ddl_test.(*testFailDBSuite).TestHalfwayCancelOperations()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/ddl/fail_db_test.go:125 +0x1adb
  github.com/pingcap/tidb/ddl_test.(*testFailDBSuite).TestHalfwayCancelOperations()

 github.com/pingcap/tidb/ddl_test.(*testFailDBSuite).TestHalfwayCancelOperations()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/ddl/fail_db_test.go:69 +0x70
  runtime.call32()
      /usr/local/go/src/runtime/asm_amd64.s:522 +0x3a
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:308 +0xc0
  github.com/pingcap/check.(*suiteRunner).forkTest.func1()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20171206051426-1c287c953996/check.go:798 +0xa04
  github.com/pingcap/check.(*suiteRunner).forkCall.func1()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20171206051426-1c287c953996/check.go:692 +0x89

Previous read at 0x00c0001d4c70 by goroutine 179:
  github.com/pingcap/tidb/domain.(*Domain).Reload()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/domain/domain.go:304 +0x151
  github.com/pingcap/tidb/domain.(*Domain).loadSchemaInLoop()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/domain/domain.go:401 +0x7f1
```

### What is changed and how it works?

* Data race error when concurrently accessing do.infoHandle in this [PR test](https://github.com/pingcap/tidb/pull/7997/files#diff-46dccda7ce1de403b0cda3709f1ce5ceR537).
* So here is the atomic pointer operation on do.infoHandle.

### Check List 

Tests 

 - Unit test

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8287)

### CODE REVIEW

File: domain/domain.go [hunk 1]
```diff
@@ -535,7 +535,8 @@ func NewDomain(store kv.Storage, ddlLease time.Duration, statsLease time.Duratio
 
 // ResetHandle resets the domain's infoschema handle. It is used for testing.
 func (do *Domain) ResetHandle(store kv.Storage) {
```
[the remark this answers is not in the document]
Comment (winkyao) [reply domain/domain.go]: Call do.Reload still has data race.

File: domain/domain.go [hunk 2]
```diff
@@ -535,7 +535,8 @@ func NewDomain(store kv.Storage, ddlLease time.Duration, statsLease time.Duratio
...
-	do.infoHandle = infoschema.NewHandle(store)
+	// `do.infoHandle` may refer to being accessed concurrently, so atomic operations are performed here.
+	atomic.StorePointer((*unsafe.Pointer)(unsafe.Pointer(&do.infoHandle)), unsafe.Pointer(infoschema.NewHandle(store)))
```
Comment (winkyao) [inline domain/domain.go]: Protect reading at do.infoHandle pointer.

Comment (ciscoxll) [reply domain/domain.go]: @winkyao Done.

File: domain/domain.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply domain/domain.go]: So why it's needed to reset the handle? @zimulala 
`do.Reload` uses mutex.Lock @winkyao

[the remark this answers is not in the document]
Comment (zimulala) [reply domain/domain.go]: It's used for testing. You can see it in "TestHalfwayCancelOperations". @tiancaiamao

[the remark this answers is not in the document]
Comment (tiancaiamao) [reply domain/domain.go]: I mean, could this line https://github.com/pingcap/tidb/pull/7997/files#diff-a597404ec5e4604d6b0d7d01c32700dbR95 replaced by `s.dom.Reload()`?

File: domain/domain.go (hunk 2 above)
Comment (crazycs520) [inline domain/domain.go]: the data race is between write and read.
but, there is only use atomic.Store but not use atomic.Get, Is it worked?

Comment (winkyao) [reply domain/domain.go]: No, need to protect reading at do.infoHandle.

Comment (ciscoxll) [reply domain/domain.go]: @winkyao @crazycs520 Done

File: domain/domain.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (zimulala) [reply domain/domain.go]: No.
But we can use other "DDL" statements instead.

File: domain/domain.go [hunk 3]
```diff
@@ -238,9 +238,16 @@ func (do *Domain) tryLoadSchemaDiffs(m *meta.Meta, usedVersion, newVersion int64
...
+// getInfoHandle atomic load infoschema.handle.
+func (do *Domain) getInfoHandle() *infoschema.Handle {
+	return (*infoschema.Handle)(atomic.LoadPointer(
+		(*unsafe.Pointer)(unsafe.Pointer(&do.infoHandle)),
+	))
+}
+
 // InfoSchema gets information schema from domain.
 func (do *Domain) InfoSchema() infoschema.InfoSchema {
-	return do.infoHandle.Get()
```
[the remark this answers is not in the document]
Comment (ciscoxll) [reply domain/domain.go]: @winkyao Done.

**Comment by winkyao** [comment]
@XuHuaiyu Will fix the union test failed, @ciscoxll you can just rerun the `/run-common-test`

**Comment by winkyao** [comment]
This PR need to be cherry-pick to 2.1 and 2.0

File: domain/domain.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (zimulala) [reply domain/domain.go]: @ciscoxll

[the remark this answers is not in the document]
Comment (winkyao) [reply domain/domain.go]: @zimulala What do you mean?

**Comment by tiancaiamao** [comment]
@zimulala @crazycs520

**Comment by ciscoxll** [comment]
@zimulala @tiancaiamao @crazycs520 PTAL

**Comment by ciscoxll** [comment]
@zimulala @crazycs520 PTAL

**Comment by ciscoxll** [comment]
@zimulala @tiancaiamao  Need to discuss it?

File: domain/domain.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ciscoxll) [reply domain/domain.go]: @zimulala @tiancaiamao Are we going to continue using `s.dom.ResetHandle`?

[the remark this answers is not in the document]
Comment (zimulala) [reply domain/domain.go]: @winkyao @ciscoxll 
I mean we can change test.  Maybe we can remove `ResetHandle`.
At most, the test results are not so good.

File: ddl/fail_db_test.go [hunk 4]
```diff
@@ -93,7 +93,8 @@ func (s *testFailDBSuite) TestHalfwayCancelOperations(c *C) {
 	c.Assert(row.GetInt64(0), DeepEquals, int64(1))
 	c.Assert(rs[0].Close(), IsNil)
 	// Reload schema.
-	s.dom.ResetHandle(s.store)
+	_, err = s.se.Execute(context.Background(), "alter table t add column t1 int")
```
Comment (winkyao) [inline ddl/fail_db_test.go]: use `alter comment` to reload schema.

File: ddl/fail_db_test.go [hunk 5]
```diff
@@ -125,7 +126,8 @@ func (s *testFailDBSuite) TestHalfwayCancelOperations(c *C) {
 	c.Assert(row.GetInt64(0), DeepEquals, int64(1))
 	c.Assert(rs[0].Close(), IsNil)
 	// Reload schema.
```
Comment (zz-jason) [inline ddl/fail_db_test.go]: should we update the code comment?

File: ddl/fail_db_test.go [hunk 6]
```diff
@@ -93,7 +93,8 @@ func (s *testFailDBSuite) TestHalfwayCancelOperations(c *C) {
...
```
Comment (zz-jason) [inline ddl/fail_db_test.go]: ditto

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**